### PR TITLE
Fix find char with line ending

### LINKED ISSUE #14361: repeat_last_motion inconsistent behaviour
### Summary

There seems to be a difference between `f<char>` or `t<char>` followed by `Alt+.` and `f<return>` or `t<return>` followd by `Alt+.`.

Repeating the motion with normal chars works as expected, but repeating the motion to `return` char does not work as expected, it selects to the char that was selected previously.

### Reproduction Steps

* Text: `Hamburgefontsiv \n Hamburgefontsiv`

### Expected

* `fb` selects `Hamb` `Alt+.` selects `Hamburgefontsiv \n Hamb`

### Not Expected
* `f<ret> selects `Hamburgefontsiv \n` but then `Alt+.` selects `Hamburgefontsiv \n Hamb`

### Platform

Linux

### Terminal Emulator

alacritty

### Installation Method

from source

### Helix Version

25.07-144-ge3fafb6b

### PR DESCRIPTION
**Pull request by drybalka**
Fixes #14361 

Simplifies and streamlines the char finding logic

### CODE REVIEW

File: helix-term/src/commands.rs [hunk 1]
```diff
@@ -1491,51 +1491,53 @@ fn extend_next_sub_word_end(cx: &mut Context) {
 // This is necessary because the one document can have different line endings inside. And we
 // cannot predict what character to find when <ret> is pressed. On the current line it can be `lf`
 // but on the next line it can be `crlf`. That's why [`find_char_impl`] cannot be applied here.
-fn find_char_line_ending(
-    cx: &mut Context,
+fn find_char_line_ending_motion(
```
Comment (archseer) [inline helix-term/src/commands.rs:1495]: Can you add some integration tests for these? You can translate the reproduction case in the linked issue

Comment (drybalka) [reply helix-term/src/commands.rs:1495]: I've added the integration tests for find_char motions and their repeats. I've covered most edge cases as they can be tricky to get right.

**Comment by drybalka** [comment]
Btw, @archseer  while working on this PR I realized that the concept of "cursor" the way it is now is not ideal and maybe even harmful. 
- From the regular helix use one may think that a cursor is a single-symbol Range, but in the code the method `cursor` "Gets the left-side position of the block cursor" of type `usize`, which means that the cursor is simply a gap between characters. Moreover, it may coincide with a range's lhs, but not rhs, which is annoying.
- The cursor size may be 1 or more chars depending on the underlying text, which is strange. A Range, in contrast, does not have this problem.
- Because of the requirement that a cursor may not be zero the following ranges are implicitly non-representable: `Range { anchor: 1, head: 0 }` (inverted single-char range) and `Range { anchor: 1, head: 1}` (zero-width range). This "invariant", however, is not reflected in the type system, which may lead to bugs. Basically one must always keep this fact in mind when coding.
- It adds cognitive overhead as a selection and a cursor have overlapping responsibilities. For example, `i`/`a` adds text around the cursor, whereas `p`/`P` pastes around the selection - and one needs to remember for each command in which way it behaves.

It would have been much cleaner if instead of a "block-cursor" there existed only selections, whereas the cursor itself had been one of the edges of a selection, a gap, same as it already is in code, basically. I don't see much benefit in upholding the following invariant "2. All ranges are at least 1 character wide, unless at the very end of the document.", which even in itself contains an exception and is not persisted in type system in any way. Are there any reasons for it other than historical or "this implicit behavior is so deep in code, that breaking it is dangerous at this point"? Because I would be interested in implementing it, even as an optional feature  `zero_width_cursor = true` or something. Or do you think this is a waste of time and you will never merge it?

**Comment by archseer** [comment]
Yes, the implementation is of course quite deliberate: the `Range` code is used for all sorts of OT operations and gap indexing is the most flexible since it allows you to mark a range of text and make transformations to it with the resulting selection still making sense. I might be misremembering I based this on CodeMirror after having issues with Vim's mark internals for tracking placeholder values in snippet.

So why don't we support empty range selections in Helix UI? Well, terminal cursors are block by default so there wasn't a good way to distinguish between a 1-char range and an empty range. With modern terminals this is of course solvable (except we still wouldn't be able to render it for selections other than the primary one), but since the kakoune/helix model operates on selections an empty range is rarely useful and more of a nuisance.

So for the time being we normalize 0-width back into 1-width selections after edits are applied. In an alternative UI (wgpu, web based etc.) we were going to experiment exposing this to the user

**Comment by archseer** [comment]
Okay, looks like I am somewhat misremembering: we were originally using gap indexing in transaction code but not on ranges which came with it's own set of bugs, and we normalized it in https://github.com/helix-editor/helix/issues/362 (and particularly https://github.com/helix-editor/helix/issues/362#issuecomment-867804971)

**Comment by drybalka** [comment]
@archseer Thank you for the prompt reply! I see in #362 you already discussed the zero-width cursor and I agree on most of the author's ideas. After some research I now also understand that showing a zero-width multi-cursor is tricky in a terminal (which is somewhat a shame) and therefore is not worth it. 

However, I still stand by the second part of my proposal, which is also the logical conclusion of the work done in #362. I propose to treat the cursor as just an either end of the main selection, and for all the commands that assume cursor to be a block to treat the main selection as the fat cursor. I am sorry if this is formulated awkwardly, but hopefully you get my point. This will make the insert/append commands to add text before or after the main selection range, instead of the "lhs/rhs edge of the grapheme to the right from the cursor gap-index". This will not only unify the command behavior (paste already behaves like that), but also simplify the code as cursor will now be simply head of the main selection, instead of the awkward:
```rust
        if self.head > self.anchor {
            prev_grapheme_boundary(text, self.head)
        } else {
            self.head
        }
```
And this is not the only place with such an awkward code.

Btw, I can make a formal proposal like in #362 if you want, I just suspect it will be thrown into the discussion bin and just be forgotten there =)

**Comment by drybalka** [comment]
In any case, my suggestions above are irrelevant for this PR, which just simply fixes a bug. I will move them into a separate discussion.

**Comment by drybalka** [comment]
@archseer and other maintainers, this is just a small ping so that this bug-fixing PR is not lost. Thank you for your work!
